### ISSUE TITLE
Could not parse version constraint ^2.0.x-dev: Invalid version string "^2.0.x-dev"

### ISSUE THREAD

**Issue by cmmartin** [kylekatarnls/update-helper#9]
**[UPDATE]**

The following issue raised for composer 1.10.11 and 1.10.12, please run:
```
composer self-update
```
to get 1.10.13 or higher
_- @kylekatarnls_

-----------------

I'm following the install instructions for Laravel valet found [here](https://laravel.com/docs/7.x/valet)

The below step fails...

```
Install Valet with Composer via composer global require laravel/valet. Make sure the ~/.composer/vendor/bin directory is in your system's "PATH".
```

with the following error...

```
  Could not load package kylekatarnls/update-helper in http://repo.packagist.org: [UnexpectedValueException] Could not parse version constraint ^2.0.x-dev: Invalid version string "^2.0.x-dev"
```

Any idea how I can resolve this?

**Comment by kattatzu** [comment]
<details> 
<summary>Please follow the version standard, composer fail at install dependencies where include this library.</summary> 

```
 [RuntimeException]
  Could not load package kylekatarnls/update-helper in http://repo.packagist.org: [UnexpectedValueException] Could not parse version constraint ^2.0.x-dev: Invalid version string "^2.0.x-dev"


Exception trace:
 () at phar:///usr/local/bin/composer/src/Composer/Repository/ComposerRepository.php:668
 Composer\Repository\ComposerRepository->createPackage() at phar:///usr/local/bin/composer/src/Composer/Repository/ComposerRepository.php:412
 Composer\Repository\ComposerRepository->whatProvides() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/Pool.php:213
 Composer\DependencyResolver\Pool->computeWhatProvides() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/Pool.php:202
 Composer\DependencyResolver\Pool->whatProvides() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:173
 Composer\DependencyResolver\RuleSetGenerator->whitelistFromPackage() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:153
 Composer\DependencyResolver\RuleSetGenerator->allowListFromPackage() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:322
 Composer\DependencyResolver\RuleSetGenerator->whitelistFromJobs() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:309
 Composer\DependencyResolver\RuleSetGenerator->allowListFromJobs() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:373
 Composer\DependencyResolver\RuleSetGenerator->getRulesFor() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/Solver.php:217
 Composer\DependencyResolver\Solver->solve() at phar:///usr/local/bin/composer/src/Composer/Installer.php:489
 Composer\Installer->doInstall() at phar:///usr/local/bin/composer/src/Composer/Installer.php:232
 Composer\Installer->run() at phar:///usr/local/bin/composer/src/Composer/Command/UpdateCommand.php:163
 Composer\Command\UpdateCommand->execute() at phar:///usr/local/bin/composer/vendor/symfony/console/Command/Command.php:245
 Symfony\Component\Console\Command\Command->run() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:835
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:185
 Symfony\Component\Console\Application->doRun() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:281
 Composer\Console\Application->doRun() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:113
 Composer\Console\Application->run() at phar:///usr/local/bin/composer/bin/composer:61
 require() at /usr/local/bin/composer:24


  [UnexpectedValueException]
  Could not parse version constraint ^2.0.x-dev: Invalid version string "^2.0.x-dev"


Exception trace:
 () at phar:///usr/local/bin/composer/vendor/composer/semver/src/VersionParser.php:488
 Composer\Semver\VersionParser->parseConstraint() at phar:///usr/local/bin/composer/vendor/composer/semver/src/VersionParser.php:260
 Composer\Semver\VersionParser->parseConstraints() at phar:///usr/local/bin/composer/src/Composer/Package/Version/VersionParser.php:29
 Composer\Package\Version\VersionParser->parseConstraints() at phar:///usr/local/bin/composer/src/Composer/Package/Loader/ArrayLoader.php:242
 Composer\Package\Loader\ArrayLoader->parseLinks() at phar:///usr/local/bin/composer/src/Composer/Package/Loader/ArrayLoader.php:121
 Composer\Package\Loader\ArrayLoader->load() at phar:///usr/local/bin/composer/src/Composer/Repository/ComposerRepository.php:659
 Composer\Repository\ComposerRepository->createPackage() at phar:///usr/local/bin/composer/src/Composer/Repository/ComposerRepository.php:412
 Composer\Repository\ComposerRepository->whatProvides() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/Pool.php:213
 Composer\DependencyResolver\Pool->computeWhatProvides() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/Pool.php:202
 Composer\DependencyResolver\Pool->whatProvides() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:173
 Composer\DependencyResolver\RuleSetGenerator->whitelistFromPackage() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:153
 Composer\DependencyResolver\RuleSetGenerator->allowListFromPackage() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:322
 Composer\DependencyResolver\RuleSetGenerator->whitelistFromJobs() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:309
 Composer\DependencyResolver\RuleSetGenerator->allowListFromJobs() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/RuleSetGenerator.php:373
 Composer\DependencyResolver\RuleSetGenerator->getRulesFor() at phar:///usr/local/bin/composer/src/Composer/DependencyResolver/Solver.php:217
 Composer\DependencyResolver\Solver->solve() at phar:///usr/local/bin/composer/src/Composer/Installer.php:489
 Composer\Installer->doInstall() at phar:///usr/local/bin/composer/src/Composer/Installer.php:232
 Composer\Installer->run() at phar:///usr/local/bin/composer/src/Composer/Command/UpdateCommand.php:163
 Composer\Command\UpdateCommand->execute() at phar:///usr/local/bin/composer/vendor/symfony/console/Command/Command.php:245
 Symfony\Component\Console\Command\Command->run() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:835
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:185
 Symfony\Component\Console\Application->doRun() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:281
 Composer\Console\Application->doRun() at phar:///usr/local/bin/composer/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at phar:///usr/local/bin/composer/src/Composer/Console/Application.php:113
 Composer\Console\Application->run() at phar:///usr/local/bin/composer/bin/composer:61
 require() at /usr/local/bin/composer:24

update [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--lock] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [--no-suggest] [--with-dependencies] [--with-all-dependencies] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--ignore-platform-reqs] [--prefer-stable] [--prefer-lowest] [-i|--interactive] [--root-reqs] [--] [<packages>]...
```

</details>

**Comment by dmitry-kulikov** [comment]
Seems to be issue with newest version of Composer 1.10.11 (2020-09-08 16:53:44). Downgrade Composer until the maintainer fixes the issue:
```sh
composer self-update 1.10.10
```

**Comment by stevebauman** [comment]
Getting this as well in Travis-CI.

**Comment by kylekatarnls** [comment]
@kattatzu This kind of version is standard and even comes from the Composer documentation itself:
https://getcomposer.org/doc/articles/versions.md#branches

And Laravel uses the same standard for it's dev versions.

First, I don't have enough steps and context such as the initial composer.json, version of Laravel, PHP version. So I can know which package requires kylekatarnls/update-helper in this situation.

If you have a Laravel version >= 5.8, you should not even have the package kylekatarnls/update-helper installed.

**Comment by stevebauman** [comment]
Hi @kylekatarnls, thanks for the quick reply.

I'm experiencing this right now in Travis-CI running tests. All of my Laravel packages are encountering the same error:

https://travis-ci.org/github/larapacks/authorization/jobs/725313146

<img width="908" alt="Screen Shot 2020-09-08 at 1 17 14 PM" src="https://user-images.githubusercontent.com/6421846/92507852-9fd3b000-f1d5-11ea-8492-993cbc4abebf.png">

Setting the composer version manually to `1.10.1` as @dmitry-kulikov mentioned, resolves the issue:

https://travis-ci.org/github/larapacks/authorization/jobs/725316652

Here's my `.travis.yml` config:

```
language: php

php:
  - 7.1
  - 7.2
  - 7.3
  - 7.4

before_script:
  - travis_retry composer self-update
  - travis_retry composer install --prefer-source --no-interaction --dev

script: ./vendor/bin/phpunit
```

And here's my `composer.json` for the package in question:

```json
{
  "name": "larapacks/authorization",
  "description": "Native Laravel Authorization.",
  "keywords": ["authorization", "laravel", "permissions", "roles", "acl"],
  "license": "MIT",
  "type": "project",
  "require": {
    "php": ">=7.0",
    "illuminate/database": "^5.5|^6.0|^7.0|^8.0"
  },
  "require-dev": {
    "orchestra/testbench": "^3.7|^4.0|^5.0",
    "phpunit/phpunit": "^7.0|^8.0|^9.0"
  },
  "autoload": {
    "psr-4": {
      "Larapacks\\Authorization\\": "src/"
    }
  },
  "autoload-dev": {
    "psr-4": {
      "Larapacks\\Authorization\\Tests\\": "tests/"
    }
  },
  "extra": {
    "laravel": {
      "providers": [
        "Larapacks\\Authorization\\AuthorizationServiceProvider"
      ]
    }
  }
}
```

Let me know if you need any further information. Thanks!

**Comment by kylekatarnls** [comment]
Composer itself use it for its own versions:
https://packagist.org/packages/composer/composer#1.10.x-dev

If we require `"composer/composer": "2.0.x-dev || ^2.0.0-dev"` that's precisely because those versions existed and were valid when we added it.

And this kind of range is widely used. Many other packages would fail to resolve the version.

The version 2 of composer is an alpha (not ready to use in production). This is very likely an error in this release.

**Comment by cmmartin** [comment]
This is happening to me using `Composer version 1.10.11 2020-09-08 16:53:44`

Please let me know if you need any more information. I'm brand new to using composer and simply did a clean install on a new machine following [these docs](https://laravel.com/docs/7.x/valet)

**Comment by kylekatarnls** [comment]
I'll keep an eye on https://github.com/composer/composer/issues/9192 and https://github.com/composer/composer/issues/9191 see and wait for feedback from composer before taking any decision.

Thanks for the report.

**Comment by stevebauman** [comment]
> If we require `"composer/composer": "2.0.x-dev || ^2.0.0-dev"` that's precisely because those versions existed and were valid when we added it.

Should we use the constraint `"composer/composer": "^1.10 || 2.0.x-dev || ^2.0.0-dev"` to prevent this from occurring? Is there a purpose for forcing the `v2` alpha version of composer?

> The version 2 of composer is an alpha (not ready to use in production). This is very likely an error in this release.

Which release? In this package, or composer?

**Comment by kylekatarnls** [comment]
This is a `require-dev`. It means this is only installed to develop kylekatarnls/update-helper, not for package that uses kylekatarnls/update-helper.

If composer fails to parse `"2.0.x-dev || ^2.0.0-dev"`, it will fail to parse `"^1.10 || 2.0.x-dev || ^2.0.0-dev"` so no, I doubt this will help.

**Comment by stevebauman** [comment]
> This is a require-dev. It means this is only installed to develop kylekatarnls/update-helper, not for package that uses kylekatarnls/update-helper.

Understood, thanks! 👍

**Comment by kylekatarnls** [comment]
For everyone here, until a fix from composer, please use composer <= 1.10.10 (neither 2+ nor 1.10.11+)

**Comment by yondifon** [comment]
Upgrade to 1.10.11. Have the same problem

**Comment by cedricve** [comment]
Upgrade to 1.10.11 helped.

**Comment by toyi** [comment]
Same error as well, `composer self-update 1.10.10` "fixed" it.

**Comment by lucasres** [comment]
Same error, but i fix with downgrade the composer version to 1.10.10. But have any solution withou downgrade composer?

**Comment by dmitry-kulikov** [comment]
It was bug in Composer, already fixed, upgrade Composer to 1.10.13, https://github.com/composer/composer/issues/9191.
